README

DISCLAIMER:

THIS IS A DEVELOPMENT VERSION OF NEXT GENERATION OF SCRATCHBOX 2,
called "2.3-series" during the development.
So, this version should be regarded as an early prototype;
if you need a stable version, please use the latest one
from the 2.2-series (currently 2.2.2).

Original "README" follows:
---------------------------------------
SCRATCHBOX 2 README

author: Lauri Leukkunen <lle@rahina.org>
date:   2008-10-20
(minor updates by Lauri T. Aarnio, 2011-03-29)



LICENSING

All files are licensed under GNU LGPL version 2.1 unless specifically 
stated otherwise in the file itself.



GENERAL NOTICE

This is no longer only development code, it is expected to work.
If it doesn't, please report bugs to the scratchbox development
mailing list (scratchbox-devel@lists.scratchbox.org)


INSTALLATION

You need git to work with sbox2, get it from http://git.or.cz/.

Clone the scratchbox 2 repository:
$ git clone git://gitorious.org/scratchbox2/scratchbox2.git
or if behind a corporate firewall, use http:
$ git clone http://git.gitorious.org/scratchbox2/scratchbox2.git

(the old git repository at freedesktop.org is not used anymore)

For x86_64 hosts Scratchbox 2 automatically builds both i386
and x86_64 versions of libsb2.so, for this reason you need
to have the necessary 32bit support libs available (libc6-dev-i386,
lib32gcc1 & co. on debian).

There are two ways to build and install SB2:

1) for Debian (and derivatives, including Ubuntu):

	$ cd scratchbox2
	$ dpkg-buildpackage -rfakeroot
	$ cd ..
	$ sudo dpkg -i libsb2*deb scratchbox2*deb

2) for non-debian systems:

	$ cd scratchbox2
	$ ./autogen.sh
	$ make install prefix=$HOME/sb2

	(Perhaps adding $HOME/sb2/bin to your PATH is also a good idea, put
	"export PATH=$PATH:$HOME/sb2/bin")

You need Qemu for cpu-transparency. As of 2009-03, qemu-0.10.1 seems to work
pretty well. In addition to that, "fakeroot" and "realpath" are required;
these are usually available in the host OS (or can be installed with the
usual system management utilities, like "apt-get")

Latest Qemu can be retrieved with: 

$ git clone git://git.savannah.nongnu.org/qemu.git
$ cd qemu
$ ./configure --prefix=$HOME/sb2 --target-list=arm-linux-user
$ make && make install

If you're targeting ARM you can get a good toolchain from CodeSourcery:
http://www.codesourcery.com/gnu_toolchains/arm/download.html

Notice! Old Scratchbox 1 toolchains don't work with SB2.

To use sb2 you need to get a rootfs for your target and put it into
some useful directory, alternatively you can copy the necessary files
into right places from your toolchain. The essential thing is that
sb2 needs FHS compliant target directory structure to work.

From a codesourcery toolchain, you can find the "seed" target rootfs contents
from arm-2008q3/arm-none-linux-gnueabi/libc, simply copy that to the buildroot.

To complete sb2 target setup:

$ mkdir $HOME/buildroot
$ cd $HOME/buildroot
$ cp -a $HOME/arm-2008q3/arm-none-linux-gnueabi/libc/{lib,etc,usr} .
$ sb2-init my_target arm-none-linux-gnueabi-gcc

That will automatically generate a working target config for the compiler
given on the command line and run sb2-build-libtool script to get you a
nice working libtool for your compiler. If the sb2-build-libtool part
fails for some reason (incorrect http proxy or something similar), you
can always run it manually later.

At this point you can simply run sb2:

$ sb2

sb2-init made this target your default if you had no prior sb2 targets
configured. To change the default use sb2-config -d your_target.

After that you can run sb2 easily from anywhere and it'll just do its magic.
Continuing the above examples, here's how to compile and run a simple
hello-world program:

$ sb2 gcc -o hello hello.c
$ sb2 ./hello
Hello, World!


MAEMO DEVELOPMENT

The Maemo SDK+ project at http://maemo-sdk.garage.maemo.org/ provides
everything you need to develop for Maemo using SB2.


DIAGNOSING PROBLEMS WITH SB2

Sometimes everything does not work as expected, so it is good to know
how to locate problems (typically problems are related to misconfigured
mapping rules). Scratchbox 2 can produce extensive logs about all
operations, but it does not do so by default. So the first step is
to enable logging: Use the "-d" or "-L <level>" options of "sb2".

For example,

$ sb2 -L info dpkg-buildpackage -rfakeroot -d

will produce a log of most important actions that were performed by
"dpkg-buildpackage" and all subprocesses started by it. The log will
be created under ~/sb2_logs/ (The above command will also print out the
exact name and location of the logfile.)

The produced logfile can be quite huge, especially when using the higher
logging levels. A logfile analyzer script, sb2-logz, can be used to extract
essentials like error and warning messages and path mapping actions from
the logged information; log level 'info' in enough for anything that
sb2-logz needs. Higher logging levels ('debug', etc) are mostly useful
only for developers of scratchbox 2.

Another useful helper tool is "sb2-show", an utility which can be used to 
test path and command mapping rules while scratchbox 2 is active.

For example,

$ sb2-show path /etc/apt

$ sb2-show -b apt-get path /etc/apt

will show how the /etc/apt database has been mapped. The latter command
uses "apt-get" as the name of the calling program (the results may
be different depending on mapping mode, name of the calling program, etc)

sb2-show can also be used to see how parameters to certain processes will
be modified:

$ sb2-show exec gcc foo.c

shows which "gcc" would be started inside the sb2 session, together with
the actual, possibly modified, arguments. The result depends on how sb2 
was configured (see notes about sb2-init above)


KNOWN ISSUES

1. Dmalloc is utterly and completely incompatible with Scratchbox 2.
2. Performance is not as good as it could be.
3. Memory leaks exist, but they have not posed significant problems
   so far.

-- 
That's all folks, enjoy!



debian/changelog
scratchbox2 (2.3.90) unstable; urgency=low

  * No changes to the SW; instead, this version can be considered 
    to be the first release candidate for SB2 version 2.4!
  * Added source of the arch.document (Scratchbox 2 Internals
    and Architecture)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 05 Sep 2012 14:47:00 +0300

scratchbox2 (2.3.63) unstable; urgency=low

  * Implemented resolution of symbolic links inside
    simulated chroots (absolute symlinks, relative
    symlinks were mostly ok already in 2.3.62)
  * obs-rpm-* modes import /dev and /proc rules
    from the rule library
  * new experimental mapping modes obs-deb-install
    and obs-deb-build; these haven't been tested yet
    and almost certainly will need some tuning.
  * New flag for rules: force_orig_path_unless_chroot
  * Library interface version = 130

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 05 Jun 2012 19:05:00 +0300

scratchbox2 (2.3.62) unstable; urgency=low

  * Initial support for simulating chroot() (ordinary files
    and directories should be OK, but symlinks are not yet
    expected to work correctly in the simulated chroot)
  * Library interface version = 129

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Fri, 01 Jun 2012 21:57:00 +0300

scratchbox2 (2.3.61) unstable; urgency=low

  * A fix to a bug which was introduced in 2.3.60
  * Code cleanups (removed dead code etc. from pathmapping subsystem)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 30 May 2012 10:37:00 +0300

scratchbox2 (2.3.60) unstable; urgency=low

  * Fixed execution of host binaries which have SUID bit,
    SGID bit or special capabilities: These are now started
    indirectly and SUID/SGID privileges are simulated.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 29 May 2012 11:09:00 +0300

scratchbox2 (2.3.59-1) unstable; urgency=low

  * merge debian packaging from master

 -- Jussi Hakala <jussi.hakala@iki.fi>  Thu, 24 May 2012 23:08:07 +0300

scratchbox2 (2.3.58-1) unstable; urgency=low

  * Bugfix: Fixed handling of target-specific LD_PRELOAD
    and LD_LIBRARY_PATH (those can now be set by e.g.
    "sb2-config setenv LD_LIBRARY_PATH ...")

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 16 May 2012 18:24:00 +0300

scratchbox2 (2.3.57-1) unstable; urgency=low

  * More enhancements to exec functionalty [from Mika Westerberg]
  * Added support for FS rule libraries [Lauri Aarnio]
  * "emulate" mode now imports rules for /dev and /sys
    from the rule library
  * A new rule for /sys/fs/selinux is supposed to hide selinux from
    the target side (in emulate mode), and that way should fix the
    problems caused by a active selinux on the host side.
  * libsb2.so is not linked with -lm anymore

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 08 May 2012 20:16:00 +0300

scratchbox2 (2.3.56-1) unstable; urgency=low

  * Various fixes and enhancements for host-compatible
    targets (i.e. when compiling for intel targets on
    an intel host) [from Mika Westerberg]
  * "emulate" mode: Fixes to /dev mapping rules (e.g.
    "echo >/dev/null" works again with "sb2 -e", and
    not only when "sb2 -eR" is used) [Lauri Aarnio]

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Fri, 27 Apr 2012 18:25:00 +0300

scratchbox2 (2.3.55-1) unstable; urgency=low

  * Added document: Scratchbox 2 Internals and Architecture
    (a .pdf file in the docs subdirectory) 

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 26 Apr 2012 16:41:00 +0300

scratchbox2 (2.3.54-1) unstable; urgency=low

  * A fix to emulate+toolchain mode [from Mika Westerberg]
  * more /dev fixes for the "emulate" mode
  * an additional trick for supporting debootrapping (see
    full comment in the "sb2" script)

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 19 Apr 2012 19:58:00 +0300

scratchbox2 (2.3.53-1) unstable; urgency=low

  * New manual page libsb2_interface(7), generated automatically
    from interface.master
  * fixes for /proc and /dev; "emulate" mode now allows simulated 
    devices to be created under /dev
  * Library interface version = 128
  * Improved checking of rules at startup, refuses to start up if
    there were errors
  * A change to mapping rule format:
    Rules which may not have a valid selector ('dir', 'prefix'
    or 'path' may be nil) need to be marked with 'optional_rule = true'
    (fs_rules: rule_file_interface_version = "105" now). Failing to
    do so will generate warnings to sb2d's log. In the future that
    will be an error which will prevent startup. 
  * + other small fixes.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Wed, 18 Apr 2012 17:25:00 +0300

scratchbox2 (2.3.52-1) unstable; urgency=low

  * Virtual permission subsystem: For the simulated root user,
    access(), euidaccess(), eaccess() and faccessat() will
    simulate the checks.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 12 Apr 2012 17:37:00 +0300

scratchbox2 (2.3.51-1) unstable; urgency=low

  * Bugfix to native static binary exec code
  * sb2-show: Removed command "execluafile" and options -x and -X
  * Moved some obsolete C and Lua files to unsupported/obsolete

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Tue, 10 Apr 2012 17:57:00 +0300

scratchbox2 (2.3.50-1) unstable; urgency=low

  * No more Lua in libsb2. Lua is now used only in sb2d.
    The runtime "core" of Scratchbox 2 operates now completely
    without the Lua interpreter. Lua is used at session setup
    time from sb2d, but after that, Lua is inactive.
  * All parts of exec postprocessing are now implemented in C,
    except one which was disabled: sbrsh is not anymore available
    as a CPU transparency method.
  * Qemu is the only supported CPU transparency method currently;
    regardless of what has been configured with sb2-init, libsb2
    assumes it is Qemu.
  * The old Lua-related files are still there, and there is 
    dead code in the C files, too. Those are now waiting for a
    bigger cleanup.

 -- Lauri T. Aarnio <lauri.aarnio@iki.fi>  Thu, 05 Apr 2012 17:51:00 +0300


include/sb2.h
/*
 * Copyright (C) 2006,2007 Lauri Leukkunen <lle@rahina.org>
 *
 * Licensed under LGPL version 2.1, see top level LICENSE file for details.
 */

#ifndef __SB2_H
#define __SB2_H

#include <stdio.h>
#include <time.h>
#include <stdarg.h>

/* WARNING!!
 * pthread functions MUST NOT be used directly in the preload library.
 * see the warning in luaif/luaif.c, also see the examples in that
 * file (how to detect if pthread library is available, etc)
*/
#include <pthread.h>

#if 0
#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>
#endif

#ifdef DMALLOC
#include <dmalloc.h>
#endif

/* "sb2context" used to be called "lua_instance", but
 * was renamed because it is now used for other purposes,
 * too, and can even be used without activating Lua system
 * (i.e. the "lua" pointer may remain NULL. See
 * functions "get_sb2context() and get_sb2context_lua())
*/
struct sb2context {
#if 0
	/* Lua: */
	lua_State *lua;
#endif

	/* general: */
	int mapping_disabled;
	int sb2context_in_use; /* used only if debug messages are active */

	/* for path mapping logic: */
	char *host_cwd;
	char *virtual_reversed_cwd;
};

/* Library interface version string:
 *
 * Originally, This version string was used to check that the lua scripts offer 
 * what the C files expect, and v.v.
 * Now, when lua is not anymore included in libsb2, it is only used
 * to identify the C interface (e.g. for sb2-show), and to locate
 * a matching preload library (see session setup code in the "sb2"
 * script)
 *
 * * Differences between "28,lta-2008-09-23" and "35,lta-2008-10-01":
 *   - sbox_get_mapping_requirements(): parameter work_dir was removed
 *   - sbox_translate_path(): as above
 * * Differences between "35,lta-2008-10-01" and "53,lta-2008-11-10"
 *   - added new functions sb.get_forced_mapmode() and sb.get_session_perm()
 * * Differences between "59,lta-2008-12-04" and "53,lta-2008-11-10"
 *   - part of rule selection logic is now implemented in C.
 * * Differences between "60,2008-12-07" and "59,lta-2008-12-04"
 *   - Added special handler for /proc => sb.procfs_mapping_request() was
 *     added to luaif.c (and mapping.lua needs it)
 *   - sbox_get_mapping_requirements() now returns four values
 * * Differences between "61" and "60,2008-12-07"
 *   - added execve_map_script_interpreter()
 * * Differences between "62" and "61"
 *   - added sb.test_if_listed_in_envvar()
 * * Differences between "63" and "62"
 *   - sb_find_exec_policy() has been removed.
 * * Differences between "65" and "63"
 *   - The Lua side of the mapping engine now returns "flags" (bitmask) 
 *     to the C code; functions used to return separate booleans.
 *   - (Additionally, due to a previous bugfix in path_exists(), forcing 
 *     the library to be upgraded is a good thing to do in any case!)
 * * Differences between "66" and "65"
 *   - LD_PRELOAD and LD_LIBRARY_PATH environment variables
 *     must be set by the exec postprocessing code (in argvenvp.lua);
 *     sb_exec.c refuses to exec the program if these two are not set.
 * * Differences between "67" and "66"
 *   - Introduced a wrapper for system() (no real changes to the
 *     interface, but this fixes a serious bug introduced by the
 *     previous change; this number is incremented to force an
 *     upgrade now!)
 * * Differences between "68" and "67"
 *   - added a wrapper for popen(). Added a new function to argvenvp.lua.
 * * Differences between "69" and "68"
 *   - minor changes to the popen() implementation.
 * * Differences between "70" and "69"
 *   - bug fixes in luaif/paths.c
 * * Differences between "71" and "70"
 *   - sockaddr_un can now handle "abstract" socket names (a linux extension)
 * * Differences between "72" and "71"
 *   - added wrapper for utimensat
 * * Differences between "74" and "72"
 *   - added many wrappers (__*_chk(), etc)
 * * Differences between "75" and "74"
 *   - added features for union directory simulation
 * * Differences between "77" and "75"
 *   - Mapping rule is not anymore relayed to the exec logic;
 *     arguments and return value lists of lua functions
 *     sb_execve_map_script_interpreter() and
 *     sb_execve_postprocess() were modified.
 * * Differences between "90" and "77"
 *   - Initial version for sb2, vrs 2.3.x
 * * New in version "91":
 *   - networking rules have been added, couple
 *     of new functions were introduced.
 * * New in version "92":
 *   - "ruletree" functions have been added.
 * * New in version "93":
 *   - added sb.get_session_dir()
 * * New in version "94":
 *   - new, simpler and better ruletree api (added catalog_set(),
 *     catalog_get(), new_string() and removed stuff which
 *     was derived from the old prototype)
 * * New in version "95":
 *   - "set_path" mapping action has been added to
 *     both Lua and C mapping engines.
 * * New in version "96":
 *   - uint32 and boolean types in ruletree + related interf.functions
 * * New in version "97":
 *   - "sblib.*" functions 
 * * New in version "98":
 *   - Lua fucntion sbox_execve_preprocess() was removed
 *     exec preprocessing code is now implemented in C)
 * * New in 99: Moved exec policy selection rules to ruletree
 * * New in 120:
 *   - New function sb.test_fn_class_match()
 *   - All mapping functions take an additional parameter
 *     (func_class)
 * * New in 121:
 *   - "fakeroot" isn't supported anymore, instead
 *     SB2's own Vperm subsystem offers similar functionality
 * * 122:
 *     sb_execve_map_script_interpreter() was removed
 * * 123:
 *     sbox_get_host_policy_ld_params() was removed
 * * 124:
 *     sb_execve_postprocess_native_executable() was re-written
 *     in C => sb_execve_postprocess() must not be called with
 *     exec_type=="native" anymore
 * * 125:
 *     not directly related to Lua/C interface: Rule tree
 *     header was changed (vrs 6).
 * * 126:
 *     sbox_map_network_addr() and sb.test_net_addr_match() 
 *     were removed.
 * * 127:
 *     exec postprocessing is completely implemented in C,
 *     sb_execve_postprocess() is not needed anymore.
 * * 128:
 *     new interface classes SYMLINK and CREAT
 * * 129:
 *     chroot() is now simulated, and information about that
 *     is passed in environment variable when the simulation is
 *     active + new interface class CHROOT
 * * 130:
 *     new flag for rules: force_orig_path_unless_chroot
*/
#define SB2_LUA_C_INTERFACE_VERSION "130"

/* This version string is used to check that init.lua offers
 * what sb2d expects, and v.v.
*/
#define SB2D_LUA_C_INTERFACE_VERSION "301"

/* get sb2context, without activating lua: */
extern struct sb2context *get_sb2context(void);
/* get sb2context, activate lua if not already done: */
extern struct sb2context *get_sb2context_lua(void);
extern void sb2context_initialize_lua(struct sb2context *sb2ctx);
extern void release_sb2context(struct sb2context *ptr);

#if 0
extern char *sb_decolonize_path(const char *path);
#endif

extern int sb_next_execve(const char *filename, char *const argv [],
			char *const envp[]);

extern int do_exec(int *result_errno_ptr, const char *exec_fn_name, const char *file,
		char *const *argv, char *const *envp);

extern time_t get_sb2_timestamp(void);

extern char *procfs_mapping_request(const char *path);

#if 0
extern void dump_lua_stack(const char *msg, lua_State *L);
#endif

extern int sb_path_exists(const char *path);

/* ------ debug/trace logging system for sb2: */
#define SB_LOGLEVEL_uninitialized (-1)
#define SB_LOGLEVEL_NONE	0
#define SB_LOGLEVEL_ERROR	1
#define SB_LOGLEVEL_WARNING	2
#define SB_LOGLEVEL_NETWORK	3
#define SB_LOGLEVEL_NOTICE	4
#define SB_LOGLEVEL_INFO	5
#define SB_LOGLEVEL_DEBUG	8
#define SB_LOGLEVEL_NOISE	9
#define SB_LOGLEVEL_NOISE2	10
#define SB_LOGLEVEL_NOISE3	11

extern void sblog_init(void);
extern void sblog_init_level_logfile_format(const char *opt_level,
	const char *opt_logfile, const char *opt_format);
extern int sblog_level_name_to_number(const char *level_str);

extern void sblog_vprintf_line_to_logfile(const char *file, int line,
	int level, const char *format, va_list ap);
extern void sblog_printf_line_to_logfile(const char *file, int line,
	int level, const char *format,...);

extern int sb_loglevel__; /* do not access directly */
extern int sb_log_initial_pid__; /* current PID will be recorded here
				  * when the logger is initialized */

#define SB_LOG_INITIALIZED() (sb_loglevel__ >= SB_LOGLEVEL_NONE)

#define SB_LOG_IS_ACTIVE(level) ((level) <= sb_loglevel__)

#define SB_LOG(level, ...) \
	do { \
		if (SB_LOG_IS_ACTIVE(level)) { \
			sblog_printf_line_to_logfile( \
				__FILE__, __LINE__, level, __VA_ARGS__); \
		} \
	} while (0)

#define LIBSB2 "libsb2.so.1"

extern int sb2_global_vars_initialized__;
extern void sb2_initialize_global_variables(void);
extern char *sbox_session_dir;
extern char *sbox_session_mode;
extern char *sbox_vperm_ids;
extern char *sbox_network_mode;
extern char *sbox_orig_ld_preload;
extern char *sbox_orig_ld_library_path;
extern char *sbox_binary_name;
extern char *sbox_exec_name;
extern char *sbox_real_binary_name;
extern char *sbox_orig_binary_name;
extern char *sbox_active_exec_policy_name;
extern char *sbox_mapping_method;

extern char *sbox_chroot_path; /* virtual path to the chroot directory. */

extern void check_pthread_library(void);

extern int pthread_library_is_available; /* flag */
extern pthread_t (*pthread_self_fnptr)(void);
extern int (*pthread_mutex_lock_fnptr)(pthread_mutex_t *mutex);
extern int (*pthread_mutex_unlock_fnptr)(pthread_mutex_t *mutex);

extern int pthread_detection_done;

extern int (*pthread_key_create_fnptr)(pthread_key_t *key,
	 void (*destructor)(void*));
extern void *(*pthread_getspecific_fnptr)(pthread_key_t key);
extern int (*pthread_setspecific_fnptr)(pthread_key_t key,
	const void *value);
extern int (*pthread_once_fnptr)(pthread_once_t *, void (*)(void));

#if 0
extern void lua_string_table_to_strvec(lua_State *l,
	int lua_stack_offs, char ***args, int new_argc);
void strvec_free(char **args);
#endif

#if 0
extern int lua_bind_sblib_functions(lua_State *l);
extern int lua_sb_log(lua_State *l);
extern int lua_sb_path_exists(lua_State *l);
extern int lua_sb_debug_messages_enabled(lua_State *l);
extern int lua_sb_isprefix(lua_State *l);
extern int lua_sb_test_path_match(lua_State *l);
extern int lua_sb_readlink(lua_State *l);
#endif


extern int test_if_str_in_colon_separated_list_from_env(
	const char *str, const char *env_var_name);

#endif
